### PR TITLE
I232 submit judges' solutions with tracking

### LINKED ISSUE #232: Provide a way to submit judges' solutions for a problem (or all problems)
**Feature Description**: 

Add a feature to read the solutions source for a problem and submit those solutions.

Submissions are under:
<short_name>/submissions

Accepted/Yes submissions are under
<short_name>/submissions/accepted

No submissions are under:  
<short_name>/submissions/run_time_error
<short_name>/submissions/security_violation
<short_name>/submissions/time_limit_exceeded
<short_name>/submissions/wrong_answer

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 

Provide the user the ability to:
- select all problems and sample submissions
- select an individual problem's submissions
- provide a way to select Yes/accepted submissions only
- provide a way to select "No accepted" submissions only
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### LINKED ISSUE #974: Enhance the Judge's Submissions submitter so non-admins can use it.
**Is your feature request related to a problem?** 
The `SubmitSubmissionsPane` should be enhanced so that it works and allows a judge or an admin to determine if the judge's submissions behaved as expected.  Currently, the judge's submissions can be submitted easily enough, but there is no way to tell what the expected results are, unless you manually look at the disposition of each run along with the submission - very tedious.
...

**Feature Description**: 
Make the `SubmitSubmissionsPane` available on the **adminsitrator** and **judge** client GUI's.  Currently, its only available by manually adding it to a client.
...

**Have you considered other ways to accomplish the same thing?** 
A manual compare by a human.
...

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 
Each submission has to be tagged with the expected outcome (easily remembered in a hash map).   A Run Listener can be added to listen for judgements and match them up with the submission and its expected result.  A status screen (JTable) can be presented that show the differences, if any, and also the CPU time used for each test case.
...

**Additional context**: 
Requested by the PacNW Judging team (Nick Wu), and, @johnbrvc  
...

### PR DESCRIPTION
**Pull request by johnbrvc**
### Description of what the PR does
Allows a judge or administrator to submit the judges' sample submissions and provides a table to compare the expected results with the actual results.  Also provides detailed timings for each test case, as well as executes all test cases, even on failure.  This is necessary for a judge to validate their submissions.

Note: Some of the code in this PR came from the last branch that @lane55 was working on.  I used most of that code, and added more to finish off the PR.  Specifically, all the _list_ related code, as well as the **QuickSubmitter** stuff was from @lane55's branch.  He had the start of a GUI but I pretty much refactored that.

### Issue which the PR addresses
Fixes #232
Fixes #974 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
OS: Windows 11
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
1) Load a contest that has judges' sample solutions - _NAC2024 Dress_ is such a contest.
2) Start the `pc2server`
3) Start up one or more _autojudges_
4) Start up an `pc2admin `client
5) Start the contest
6) Navigate to the **Run->Submit Samples** tab.
7) Open the tab
8) You can choose to **Filter** some submissions by pressing the filter button and selecting what it is you're looking for, or just submit all the runs.  For _NAC 2024 Dress_ there were 68 runs, and it took about 15 mins for a single _autojudge_ to judge them.
9) For this example, I chose to submit just the sample solutions for the first problem.
image

10) Press the **Submit** button and then answer **Yes** to the "**_Submit XXX judge sample submissions?_**"
11) You should see the results appear as they submissions are submitted and judged:
image

12) Results are color-coded.  Green means the PC2 autojudged run matched what the human judge who wrote the submission expected.  Red mean it did not match what was expected.  Yellow means it could not be determined.  Note in the sample below, **Run Id 318** is yellow because the "**Expected**" result is "**rejected**", which is not a valid CLICS judgment.  The **Status** column indicates PC2 judged it as time limit exceeded:
image

13) If you click on the link in the "**Max Time(s)**" column, a window will pop up showing the results of each test case.  For example, if we click on **Run ID 318**'s time of **1.080s**, the following window appears:
image

14) The **Result** column shows the autojudge's disposition of the test case.  You can also click on the links in the **Judge's Data** and **Judge's Output** column to display the contents of the specific judges test data.

### CODE REVIEW

File: src/edu/csus/ecs/pc2/core/InternalController.java [hunk 1]
```diff
@@ -527,45 +530,85 @@ protected void sendToServersAndAdmins(Packet packet) {
...
     public void submitJudgeRun(Problem problem, Language language, String filename, SerializedFile[] otherFiles) throws Exception {
         submitJudgeRun(problem, language, filename, otherFiles, 0, 0);
     }
-    
+
+    /**
+     * {@inheritDoc}
+     * Calling this method is equivalent to calling
+     * {@link #submitJudgeRun(Problem, Language, String, SerializedFile[], long, long)}
+     * with zeroes as the last two parameters.
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/core/InternalController.java:546]: This does not match the actual thing called.

File: src/edu/csus/ecs/pc2/list/ListUtilities.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+/**
+ * A set of utilities that operate on lists.
+ *
+ * @author Douglas A. Lane <pc2@ecs.csus.edu>
+ *
+ */
+public class ListUtilities {
+
+	/**
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/list/ListUtilities.java:24]: this is indented wrong

File: src/edu/csus/ecs/pc2/list/ListUtilities.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	 *
+	 * Example files under config\sumit\submissions
+	 *
+	 * @param mycontest
+	 * @param directoryName CDP config/ dir or CDP base directory
+	 * @return list of judges sample submissions
+	 */
+	// TODO REFACTOR remove getAllCDPsubmissionFileNames from QuickSubmitter
+	public static List<File> getAllJudgeSampleSubmissionFilenamesFromCDP(IInternalContest mycontest, String directoryName) {
+
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/list/ListUtilities.java:93]: this is indented wrong

File: src/edu/csus/ecs/pc2/list/ListUtilities.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	        }
+	    }
+	    types.sort(null);
+	    return(types);
+	}
+
+	/**
+	 * Find files (in CDP) that contain one of the Problems in the list.
+	 *
+	 */
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/list/ListUtilities.java:167]: this is indented wrong

File: src/edu/csus/ecs/pc2/list/SubmissionSampleLocation.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        { CLICS_SUBMISSION_LOCATION_CE, CLICSJudgementType.CLICS_BIG5.CE.toString() }
+    };
+
+	private String title;
+	private String shortDirectoryName;
+	private String clicsAcronym;
+
+	/**
+	 * A judges solution name and location.
+	 *
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/list/SubmissionSampleLocation.java:34]: this is indented wrong

File: src/edu/csus/ecs/pc2/list/SubmissionSolutionList.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		createSubmissionDirectoryMap();
+		loadCDPSampleSubmissionList(mycontest, cdpPath);
+	}
+
+	private void createSubmissionDirectoryMap() {
+        // Create quick access to look up acronym for folder
+        for(String [] subFolderRow : SubmissionSampleLocation.CLICS_SUBMISSION_TO_ACRONYM) {
+            submissionDirectoryToAcronym.put(subFolderRow[0], subFolderRow[1]);
+        }
+	}
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/list/SubmissionSolutionList.java:43]: this is indented wrong

File: src/edu/csus/ecs/pc2/ui/SubmitSampleRunsPane.java [hunk 7]
```diff
@@ -0,0 +1,1391 @@
...
+
+	public SubmitSampleRunsPane() {
+		super();
+		setLayout(new BorderLayout());
+
+        add(getMessagePanel(), BorderLayout.NORTH);
+        add(getCenterPanel(), BorderLayout.CENTER);
+        add(getBottomPanel(), BorderLayout.SOUTH);
+
+	}
```
Comment (troy2914) [inline src/edu/csus/ecs/pc2/ui/SubmitSampleRunsPane.java:166]: indented wrong

**Comment by johnbrvc** [comment]
So, before someone complains that the CI did not complete successfully, I am aware.  These are JUnits that @lane55 wrote for the original code HE wrote, but I have since modified it.  I will address these.

**Review by troy2914 — COMMENTED** [review]
code review done,  I think you need to implement some things into MockController which caused the failure of this build.

some small corrections noted.

File: src/edu/csus/ecs/pc2/core/InternalController.java (hunk 1 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/core/InternalController.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/list/ListUtilities.java (hunk 2 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/list/ListUtilities.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/list/ListUtilities.java (hunk 3 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/list/ListUtilities.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/list/ListUtilities.java (hunk 4 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/list/ListUtilities.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/list/SubmissionSampleLocation.java (hunk 5 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/list/SubmissionSampleLocation.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/list/SubmissionSolutionList.java (hunk 6 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/list/SubmissionSolutionList.java]: Fixed and pushed. There were residual tabs from the original code.

File: src/edu/csus/ecs/pc2/ui/SubmitSampleRunsPane.java (hunk 7 above)
Comment (johnbrvc) [reply src/edu/csus/ecs/pc2/ui/SubmitSampleRunsPane.java]: Fixed and pushed. There were residual tabs from the original code.

**Review by troy2914 — COMMENTED** [review]
changes look good to me,  but build is still failing...

**Review by troy2914 — APPROVED** [review]
Err, the x for the build I saw was old apparently.

**Review by kkarakas — APPROVED** [review]
Read through the changes, nothing stood out as blatantly wrong.

**Pull request merged**